### PR TITLE
PMUX-1481 dx create_cohort

### PR DESCRIPTION
**Pull request by Niranjanpandeshwar**
Hello All,
This PR adds a new feature `create_cohort` to the suite of dx commands. The input to this is a record of type dataset or CohortBrowser(Some exceptions do exist which are handled with error handlers).
Jenkins test for this feature pass on both py2 and py3.
py2: [https://jenkins-vstg.internal.dnanexus.com/job/dx-toolkit/job/dxpy-master-integration-tests-on-staging-16.04-python2/2337/](https://jenkins-vstg.internal.dnanexus.com/job/dx-toolkit/job/dxpy-master-integration-tests-on-staging-16.04-python2/2337/)
py3: [https://jenkins-vstg.internal.dnanexus.com/job/dx-toolkit/job/dxpy-branch-integration-tests-on-staging-20.04-python3/526/](https://jenkins-vstg.internal.dnanexus.com/job/dx-toolkit/job/dxpy-branch-integration-tests-on-staging-20.04-python3/526/)

### CODE REVIEW

**Review by jkotrs — APPROVED** [review]
Minor comment around tests. Thanks!

File: src/python/dxpy/cli/dataset_utilities.py [hunk 1]
```diff
@@ -1047,6 +1058,199 @@ def print_fields(fields):
...
+    #### Reading input sample ids from file or command line ####
+    samples=[]
+    # from file
+    if args.cohort_ids_file:
+        with open(args.cohort_ids_file,"r") as infile:
+            for line in infile:
+                samples.append(line.strip("\n"))
+    # from string
+    if args.cohort_ids:
+        samples = args.cohort_ids.split(",")
```
Comment (durae) [inline src/python/dxpy/cli/dataset_utilities.py:1203]: Will it be documented that there cannot be any spaces between sample ids?

File: src/python/dxpy/cli/dataset_utilities.py [hunk 2]
```diff
@@ -1047,6 +1058,199 @@ def print_fields(fields):
...
+    
+    #### Validate the input cohort IDs ####
+    # Get the table/entity and field/column of the dataset from the descriptor
+    rec_descriptor = DXDataset(resp["dataset"], project=resp["datasetRecordProject"]).get_descriptor()
+
+    try:
+        validate_cohort_ids(rec_descriptor,dataset_project,resp,samples)
```
Comment (durae) [inline src/python/dxpy/cli/dataset_utilities.py:1210]: Can you keep spaces between args? It is in multiple places in this PR.

File: src/python/dxpy/dx_extract_utils/cohort_filter_payload.py [hunk 3]
```diff
@@ -0,0 +1,141 @@
+import copy
+
+
+def generate_pheno_filter(values, entity, field, filters):
+    filters = copy.deepcopy(filters)
```
Comment (durae) [inline src/python/dxpy/dx_extract_utils/cohort_filter_payload.py:5]: Just heads up that deepcopy is very slow for more complicated filters. Maybe better idea is to create a new filters hash and copy over things that you need will have in filters.

File: src/python/dxpy/dx_extract_utils/cohort_filter_payload.py [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+    return combined
+
+
+def cohort_final_payload(name, folder, project, databases, dataset, filters, sql, base_sql=None, combined=None):
+    details = {
+        "databases": databases,
+        "dataset": {"$dnanexus_link": dataset},
+        "description": "",
+        "filters": filters,
+        "schema": "create_cohort_schema",
```
Comment (durae) [inline src/python/dxpy/dx_extract_utils/cohort_filter_payload.py:121]: ?
Should this be schema copied from dataset details?

File: src/python/dxpy/dx_extract_utils/cohort_filter_payload.py [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+        "sql": sql,
+        "version": "3.0",
+    }
+    if base_sql is not None:
+        details["baseSql"] = base_sql
+    if combined is not None:
+        details["combined"] = cohort_combined_payload(combined)
```
Comment (durae) [inline src/python/dxpy/dx_extract_utils/cohort_filter_payload.py:128]: Why this combined cannot be just set and needs to be re-created?

**Review by durae — COMMENTED** [review]
I didn't check tests, but here are my comments for the rest of the code.

File: src/python/dxpy/cli/dataset_utilities.py (hunk 1 above)
Comment (Niranjanpandeshwar) [reply src/python/dxpy/cli/dataset_utilities.py:1203]: Good point. I think we should add support for spaces between sample IDs.

File: src/python/dxpy/dx_extract_utils/cohort_filter_payload.py (hunk 4 above)
Comment (Niranjanpandeshwar) [reply src/python/dxpy/dx_extract_utils/cohort_filter_payload.py:121]: Correct, will update.

File: src/python/dxpy/dx_extract_utils/cohort_filter_payload.py (hunk 5 above)
Comment (Niranjanpandeshwar) [reply src/python/dxpy/dx_extract_utils/cohort_filter_payload.py:128]: visualize response looks like:
"combined": {
    "logic": "INTERSECT",
    "source": [
      {
        "id": "record-xxxx",
        "project": "project-xxxx",
        "recordName": "some_name",
        "name": "some_name"
      },
      {
        "id": "record-yyyy",
        "project": "project-yyyy",
        "recordName": "name_2",
        "name": "name_2"
      }
    ]
  }
  
  Based on some manual tests, it seems like dx new command does not respect this format and it requires dnanexus_link
  "$dnanexus_link": {
                        "id": "record-GYXGv700vGPpz41VkQXZpX47",
                        "project": "project-G9j1pX00vGPzF2XQ7843k2Jq"
                    }

Comment (durae) [reply src/python/dxpy/dx_extract_utils/cohort_filter_payload.py:128]: Yes, you're right. Thanks! (I thought this is coming from record details and not from visualize.)

File: src/python/dxpy/cli/dataset_utilities.py [hunk 6]
```diff
@@ -1047,6 +1058,199 @@ def print_fields(fields):
...
+    pass
+
+def validate_cohort_ids(descriptor,project,resp,ids):
+    # Usually the name of the table
+    entity_name = descriptor.model["global_primary_key"]["entity"]
+    # The name of the column or field in the table
+    field_name = descriptor.model["global_primary_key"]["field"] 
+
+    # Prepare a payload to find entries matching the input ids in the dataset
+    table_column_name = "{}${}".format(entity_name, field_name)
```
[the remark this answers is not in the document]
Comment (Niranjanpandeshwar) [reply src/python/dxpy/cli/dataset_utilities.py:1095]: Should have been `entity_field_name`. Will change.

File: src/python/dxpy/dx_extract_utils/cohort_filter_payload.py (hunk 3 above)
Comment (Niranjanpandeshwar) [reply src/python/dxpy/dx_extract_utils/cohort_filter_payload.py:5]: We don't need to copy this dict. Will just delete this line.

**Review by ksoni — COMMENTED** [review]
Looks good to me, thanks @Niranjanpandeshwar

**Review by durae — COMMENTED** [review]
LGTM

**Pull request merged**